File: cpp/0150逆波兰表达式求值.cpp
#include<iostream>
#include<vector>
#include<stack>

using namespace std;

int GetResult(char opertator,int a,int b)
{
    int res;
    switch (opertator)
    {
        case '+':
            res = a + b;
            break;
        case '-':
            res = a - b;
            break;
        case '*':
            res = a * b;
            break;
        case '/':
            res = a / b;
            break;
        
    }
    return res;
}

int evalRPN(vector<string> &tokens)
{
    stack<int> st;
    int  a, b;
    for (string s : tokens)
    {
        if (s == "+" || s == "-" || s == "*" || s == "/")
        {
            a = st.top();
            st.pop();
            b = st.top();
            st.pop();
            st.push(GetResult(s[0], b, a));
        }
        else
            st.push(stoi(s));
    }

    return st.top();
}

int main()
{
    vector<string> t = {"2", "1", "+", "3", "*" };

    evalRPN(t);
}

File: python/main/0150逆波兰表达式求值.py

def evalRPN(tokens: list) -> int:
    def _add(a,b):
        return a+b
    def _sub(a,b):
        return a-b
    def _mul(a,b):
        return a*b
    

    memo={
        "+":_add,
        "-":_sub,
        "*":_mul,
        "/":lambda x,y:int(x/y) #注意运算符“//”是地板除，负数和题目要求向0截断不符
        }

    st=[]

    for c in tokens:
        if(c in memo.keys()):
            a=st.pop()
            b=st.pop()
            st.append(memo[c](int(b),int(a)))
        else:
            st.append(c)

    return int(st[0])

print(evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]))
